### PR TITLE
Feature/drone state

### LINKED ISSUE #21: Implement DroneState
**Is your feature request related to a problem? Please describe.**
The is no wrapper for the actual or expected state of the drone.

**Describe the solution you'd like**
There is a description in the [architecture documentation](https://treeandsea.github.io/DroneController/). But here is a brief summary: It contains several optional three dimensional vectors

- position
- translation velocity
- angular velocity
- translation acceleration
- angular acceleration

It should have a method for returning all of them in a dictionary.

**Describe alternatives you've considered**
An alternative would be to pass the values individually but this is very inefficient and error prone.

**Additional context**
This should be used for the actual drone state as well as for the expected drone state.

### PR DESCRIPTION
**Pull request by Segelzwerg**
Adds an implementation of the drone state wrapper

closes #21

### CODE REVIEW

File: src/drone_controller/exception/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class DroneControllerError(Exception):
```
Comment (Segelzwerg) [inline src/drone_controller/exception/exceptions.py]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

**Comment by Segelzwerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- src/drone_controller/input_layer/drone_state.py  2
- src/drone_controller/exception/exceptions.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178&bid=17072432)

File: src/drone_controller/exception/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+    """
```
Comment (Segelzwerg) [inline src/drone_controller/exception/exceptions.py]: Codacy Issue found: [Multi-line docstring summary should start at the first line](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: test/drone_controller/input_layer/test_drone_state.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+from pytest import raises
```
Comment (Segelzwerg) [inline test/drone_controller/input_layer/test_drone_state.py]: Codacy Issue found: [Unable to import 'pytest'](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: test/drone_controller/input_layer/test_drone_state.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        "Position": position,
+        "Velocity": velocity,
+        "Angular Velocity": velocity_ang,
+        "Acceleration": acceleration,
+        "Angular Acceleration": acceleration_ang
+    }
+
+    state_dict = drone_state.state
+
+    assert state_dict == expected_dict
```
Comment (Segelzwerg) [inline test/drone_controller/input_layer/test_drone_state.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: src/drone_controller/input_layer/drone_state.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        if not len(parameter) == 3:
+            raise DroneStateError('DroneState.create()',
+                                  (f'The input {1} is invalid {2}',
+                                   name,
+                                   parameter))
+
+
+class DroneStateError(DroneControllerError):
+    """
```
Comment (Segelzwerg) [inline src/drone_controller/input_layer/drone_state.py]: Codacy Issue found: [Multi-line docstring summary should start at the first line](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: src/drone_controller/input_layer/drone_state.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+# a wrapper class needs more parameter
+from src.drone_controller.exception.exceptions import DroneControllerError
+
+
+class DroneState:
+    """
+    Wrapper class for the state of the drone.
+    """
+
+    def __init__(self, position, velocity, velocity_ang, acceleration, acceleration_ang):
```
Comment (Segelzwerg) [inline src/drone_controller/input_layer/drone_state.py]: Codacy Issue found: [Missing docstring in __init__](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: test/drone_controller/input_layer/test_drone_state.py (hunk 3 above)
Comment (Segelzwerg) [inline test/drone_controller/input_layer/test_drone_state.py]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: src/drone_controller/input_layer/drone_state.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+# pylint: disable=too-many-arguments
```
Comment (Segelzwerg) [inline src/drone_controller/input_layer/drone_state.py]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: src/drone_controller/input_layer/drone_state.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                                   name,
+                                   parameter))
+
+
+class DroneStateError(DroneControllerError):
+    """
+    Invalid drone state.
+    """
+
+    def __init__(self, expression, message):
```
Comment (Segelzwerg) [inline src/drone_controller/input_layer/drone_state.py]: Codacy Issue found: [Missing docstring in __init__](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: test/drone_controller/input_layer/test_drone_state.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        acceleration_ang = [1, 1, 1]
+        drone_state = DroneState.create(position, velocity, velocity_ang, acceleration,
+                                        acceleration_ang)
+
+        # pylint: disable=unused-variable
+        state = drone_state.state
+
+
+def test_three_dim_vel_validation():
+    """
```
Comment (Segelzwerg) [inline test/drone_controller/input_layer/test_drone_state.py]: Codacy Issue found: [First word of the first line should be properly capitalized ('Test', not 'test')](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: test/drone_controller/input_layer/test_drone_state.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        position = [1, 1, 1]
+        velocity = [1, 1]
+        velocity_ang = [1, 1, 1]
+        acceleration = [1, 1, 1]
+        acceleration_ang = [1, 1, 1]
+        drone_state = DroneState.create(position, velocity, velocity_ang, acceleration,
+                                        acceleration_ang)
+
+        # pylint: disable=unused-variable
+        state = drone_state.state
```
Comment (Segelzwerg) [inline test/drone_controller/input_layer/test_drone_state.py]: Codacy Issue found: [local variable 'state' is assigned to but never used](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178)

File: src/drone_controller/input_layer/drone_state.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        return {
+            "Position": self._position,
+            "Velocity": self._velocity,
+            "Angular Velocity": self._velocity_ang,
+            "Acceleration": self._acceleration,
+            "Angular Acceleration": self._acceleration_ang
+        }
+
+    @classmethod
+    def create(cls, position, velocity, velocity_ang, acceleration, acceleration_ang):
```
Comment (TreeKid) [inline src/drone_controller/input_layer/drone_state.py]: Conventionally the \_\_init\_\_ method is used to create a class.
I see that you want to try out classmethods, however with this you are programming a confusing redundancy.

File: src/drone_controller/input_layer/drone_state.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self, position, velocity, velocity_ang, acceleration, acceleration_ang):
+        self._position = position
+        self._velocity = velocity
+        self._velocity_ang = velocity_ang
+        self._acceleration = acceleration
+        self._acceleration_ang = acceleration_ang
+
+    @property
+    def state(self):
```
Comment (TreeKid) [inline src/drone_controller/input_layer/drone_state.py]: maybe rename in state_dict, to make the purpose clear.

File: src/drone_controller/input_layer/drone_state.py (hunk 11 above)
Comment (Segelzwerg) [reply src/drone_controller/input_layer/drone_state.py]: yes that's true will change that

**Comment by Segelzwerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- src/drone_controller/input_layer/drone_state.py  2
- src/drone_controller/exception/exceptions.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178&bid=17072432)

**Comment by Segelzwerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- src/drone_controller/input_layer/drone_state.py  2
- src/drone_controller/exception/exceptions.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178&bid=17072432)

**Comment by Segelzwerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/drone_controller/input_layer/drone_state.py  2
- src/drone_controller/exception/exceptions.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/treeandsea/DroneController/pullRequest?prid=5205178&bid=17072432)

**Pull request merged**